File: lib/src/tray_manager.dart
import 'package:nativeapi/src/ffi/bindings.dart';
import 'package:nativeapi/src/ffi/bindings_generated.dart';

import 'tray.dart';

/// A class for managing trays.
///
/// This class provides methods for getting all trays and creating a new tray.
/// It is a singleton class.
class TrayManager {
  TrayManager._();

  /// The singleton instance of the TrayManager.
  static final TrayManager instance = TrayManager._();

  /// The native API bindings.
  NativeApiBindings get _bindings => nativeApiBindings;

  /// Create a new tray.
  ///
  /// Returns the new tray.
  Tray create() {
    final nativeTrayId = _bindings.tray_manager_create();
    return Tray(id: nativeTrayId);
  }

  /// Get a tray by its ID.
  ///
  /// Returns the tray.
  Tray get(int id) {
    return Tray(id: id);
  }

  /// Get all trays.
  ///
  /// Returns a list of all trays.
  List<Tray> getAll() {
    return [];
  }
}

File: src/tray_manager.cpp
#include <string.h>
#include <iostream>

#include "libnativeapi/include/nativeapi.h"
#include "tray_manager.h"

using namespace nativeapi;

TrayManager g_tray_manager = TrayManager();

FFI_PLUGIN_EXPORT
struct NativeTrayIDList tray_manager_get_all() {
  auto trays = g_tray_manager.GetAll();
  NativeTrayIDList native_tray_id_list;
  native_tray_id_list.count = trays.size();
  for (size_t i = 0; i < trays.size(); i++) {
    native_tray_id_list.ids[i] = trays[i]->id;
  }
  return native_tray_id_list;
}

FFI_PLUGIN_EXPORT
long tray_manager_create() {
  auto tray = g_tray_manager.Create();
  return tray->id;
}

FFI_PLUGIN_EXPORT
void tray_manager_start_listening() {
  // TODO: Implement
}

FFI_PLUGIN_EXPORT
void tray_manager_stop_listening() {
  // TODO: Implement
}
